extensions/amp-story/1.0/amp-story-background-blur.js

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {setStyles} from '#core/dom/style';
+
+/** @const {number} */
+const CANVAS_SIZE = 3;
+
+/** @const {number} */
+const DURATION_MS = 200;
+
+export class AmpStoryBackgroundBlur {
+  /**
+   * @param {!Window} win
+   * @param {!Element} element
+   */
+  constructor(win, element) {
+    /** @private @const {!Element} */
+    this.element_ = element;
+
+    /** @private @const {!Window} */
+    this.win_ = win;
+
+    /** @private @const {!Element} */
+    this.canvas_ = null;
+  }
+
+  /**
+   * Setup canvas and attach it to the document.
+   */
+  attach() {
+    this.canvas_ = document.createElement('canvas');
+    this.canvas_.width = this.canvas_.height = CANVAS_SIZE;
+    setStyles(this.canvas_, {
+      width: '100%',
+      height: '100%',
+      position: 'absolute',
+      left: 0,
+      top: 0,
+    });
+    this.element_.appendChild(this.canvas_);
+  }
+
+  /**
+   * Update the background to the specified page's background.
+   * @param {!Element} pageElement
+   */
+  update(pageElement) {
+    const fillElement = this.getBackgroundElement_(pageElement);
+    this.animate_(fillElement);
+  }
+
+  /**
+   * Animated background transition.
+   * @private
+   * @param {?Element} fillElement
+   */
+  animate_(fillElement) {
+    const context = this.canvas_.getContext('2d');
+    let startTime;
+    const nextFrame = (currTime) => {
+      if (!startTime) {
+        startTime = currTime;
+      }
+      const elapsed = currTime - startTime;
+      if (elapsed < DURATION_MS) {
+        const easing = 1 - Math.pow(1 - elapsed / DURATION_MS, 2);
+        context.globalAlpha = easing;
+        if (fillElement) {
+          context.drawImage(fillElement, 0, 0, CANVAS_SIZE, CANVAS_SIZE);
+        } else {
+          context.fillRect(0, 0, CANVAS_SIZE, CANVAS_SIZE);
+        }
+        requestAnimationFrame(nextFrame);
+      }
+    };
+    requestAnimationFrame(nextFrame);
+  }
+
+  /**
+   * Get active page's background element.
+   * @private
+   * @param {!Element} pageElement
+   * @return {?Element} An img element with template=fill or null.
+   */
+  getBackgroundElement_(pageElement) {
+    return pageElement.querySelector(
+      '[template="fill"]:not(.i-amphtml-hidden-by-media-query) img'
```

Comment:
I wonder if we should add a second CSS query as a fallback, e.g. `[template=fill] img, amp-story-grid-layer img`
If there's a template=fill it's obv a better pick, but if it's not found whatever img we can find might be good?
Some stories don't have a `template=fill`, like https://wp.stories.google/stories/web-stories-sharing-on-social-media/

Reply:
Looks pretty good in that example. I don't mind that the videos have a black background. It looks cinematic.
https://user-images.githubusercontent.com/3860311/123274485-8df79b00-d4d1-11eb-95ac-d3d5629d91ee.mp4

This story is an interesting use-case, It's getting the color from the logo. It looks branded, which is really cool but I need to add background protection. This reminds me that the mocks have a .7 overlay.
<img width="1268" alt="Screen Shot 2021-06-24 at 9 49 35 AM" src="https://user-images.githubusercontent.com/3860311/123274696-be3f3980-d4d1-11eb-8aca-4b44c0a2a557.png">